assets/js/base/hooks/use-store-cart-item.js

```diff
@@ -22,32 +24,60 @@ import { useStoreCart } from './use-store-cart';
  */
 export const useStoreCartItem = ( cartItemKey ) => {
 	const { cartItems, cartIsLoading } = useStoreCart();
-	const cartItem = cartItems.filter( ( item ) => item.key === cartItemKey );
-
-	const results = useSelect(
-		( select, { dispatch } ) => {
+	const [ cartItem, setCartItem ] = useState( {
+		key: '',
+		isLoading: true,
+		cartData: {},
+		quantity: 0,
+		isPending: false,
+		changeQuantity: () => void null,
+		removeItem: () => void null,
+	} );
+	// Store quantity in hook state. This is used to keep the UI
+	// updated while server request is updated.
+	const [ quantity, changeQuantity ] = useState( cartItem.quantity );
+	const [ debouncedQuantity ] = useDebounce( quantity, 400 );
+	const isPending = useSelect(
+		( select ) => {
 			const store = select( storeKey );
-			const isPending = store.isItemQuantityPending( cartItemKey );
-			const { removeItemFromCart, changeCartItemQuantity } = dispatch(
-				storeKey
-			);
-
-			return {
-				isPending,
-				changeQuantity: ( newQuantity ) => {
-					changeCartItemQuantity( cartItemKey, newQuantity );
-				},
-				removeItem: () => {
-					removeItemFromCart( cartItemKey );
-				},
-			};
+			return store.isItemQuantityPending( cartItemKey );
 		},
 		[ cartItemKey ]
 	);
+	useEffect( () => {
+		if ( ! cartIsLoading ) {
+			const foundCartItem = cartItems.find(
+				( item ) => item.key === cartItemKey
+			);
+			if ( foundCartItem ) {
+				setCartItem( foundCartItem );
+			}
+		}
+	}, [ cartItems, cartIsLoading, cartItemKey ] );
```

Comment:
We need to update cartItem every time `cartItems` change because the quantities will be different in the state. So basically whenever debouncedQuantity triggers an update to the state and `cartItems` change, then the internal cartItem here will change as well.